include/circt/Dialect/FIRRTL/FIRRTLDeclarations.td

```diff
@@ -199,22 +213,24 @@ def NodeOp : ReferableDeclOp<"node",
 
   let arguments = (ins PassiveType:$input, StrAttr:$name,
                        AnnotationArrayAttr:$annotations,
-                       OptionalAttr<SymbolNameAttr>:$inner_sym);
+                       OptionalAttr<SymbolNameAttr>:$inner_sym,
+                       NameKindAttr:$nameKind);
   let results = (outs FIRRTLType:$result);
 
   let builders = [
-    OpBuilder<(ins "::mlir::Type":$elementType,
-                   "::mlir::Value":$input,
-                   CArg<"StringRef", "{}">:$name,
-                   CArg<"ArrayRef<Attribute>","{}">:$annotations,
-                   CArg<"StringAttr", "StringAttr()">:$inner_sym), [{
+    OpBuilder<(ins "::mlir::Type":$elementType, "::mlir::Value":$input,
+                  CArg<"StringRef", "{}">:$name,
+                  CArg<"ArrayAttr", "ArrayAttr()">:$annotations,
+                  CArg<"StringAttr", "StringAttr()">:$inner_sym), [{
+      if(!annotations)
+        annotations = $_builder.getArrayAttr({});
       return build($_builder, $_state, elementType, input, name,
-                   $_builder.getArrayAttr(annotations), inner_sym);
+                   annotations, inner_sym, inferNameKind(name));
```

Comment:
I'm not sure this is a good idea for the builders.  infer should only be run by builders called from the parser.  Putting it here encourages random other code to use it.